be/src/vec/exec/scan/new_odbc_scanner.cpp

```diff
@@ -0,0 +1,207 @@
+    }
+
+    SCOPED_CONSUME_MEM_TRACKER(_mem_tracker);
+    // get tuple desc
+    _tuple_desc = state->desc_tbl().get_tuple_descriptor(_tuple_id);
+
+    if (nullptr == _tuple_desc) {
+        return Status::InternalError("Failed to get tuple descriptor.");
+    }
+
+    _odbc_param.connect_string = std::move(_connect_string);
+    _odbc_param.query_string = std::move(_query_string);
+    _odbc_param.tuple_desc = _tuple_desc;
+
+    _odbc_scanner.reset(new (std::nothrow) ODBCConnector(_odbc_param));
+
+    if (_odbc_scanner == nullptr) {
+        return Status::InternalError("new a odbc scanner failed.");
+    }
+
+    _text_converter.reset(new (std::nothrow) TextConverter('\\'));
+
+    if (_text_converter == nullptr) {
+        return Status::InternalError("new a text convertor failed.");
+    }
+
+    _is_init = true;
+
+    return Status::OK();
+}
+
+Status NewOdbcScanner::open(RuntimeState* state) {
+    VLOG_CRITICAL << NEW_SCANNER_TYPE << "::open";
+
+    if (nullptr == state) {
+        return Status::InternalError("input pointer is null.");
+    }
+
+    if (!_is_init) {
+        return Status::InternalError("used before initialize.");
+    }
+
+    RETURN_IF_CANCELLED(state);
+    RETURN_IF_ERROR(VScanner::open(state));
+    SCOPED_CONSUME_MEM_TRACKER(_mem_tracker);
+    RETURN_IF_ERROR(_odbc_scanner->open());
+    RETURN_IF_ERROR(_odbc_scanner->query());
+    // check materialize slot num
+
+    return Status::OK();
+}
+
+Status NewOdbcScanner::_get_block_impl(RuntimeState* state, Block* block, bool* eof) {
+    VLOG_CRITICAL << NEW_SCANNER_TYPE << "::_get_block_impl";
+
+    if (nullptr == state || nullptr == block || nullptr == eof) {
+        return Status::InternalError("input is NULL pointer");
+    }
+
+    if (!_is_init) {
+        return Status::InternalError("used before initialize.");
+    }
+    RETURN_IF_CANCELLED(state);
+
+    auto column_size = _tuple_desc->slots().size();
+    std::vector<MutableColumnPtr> columns(column_size);
+
+    bool mem_reuse = block->mem_reuse();
+    // only empty block should be here
+    DCHECK(block->rows() == 0);
+
+    // Indicates whether there are more rows to process. Set in _odbc_scanner.next().
+    bool odbc_eof = false;
+
+    do {
+        RETURN_IF_CANCELLED(state);
+
+        columns.resize(column_size);
+        for (auto i = 0; i < column_size; i++) {
+            if (mem_reuse) {
+                columns[i] = std::move(*block->get_by_position(i).column).mutate();
+            } else {
+                columns[i] = _tuple_desc->slots()[i]->get_empty_mutable_column();
+            }
+        }
+
+        for (int row_index = 0; true; row_index++) {
+            // block is full, break
+            if (state->batch_size() <= columns[0]->size()) {
+                break;
+            }
+
+            RETURN_IF_ERROR(_odbc_scanner->get_next_row(&odbc_eof));
+
+            if (odbc_eof) {
+                *eof = true;
```

Comment:
You should make sure when eof is set to true, the block you return must be empty.
Here, if `odbc_eof` is true, there may still has data in `block`.

You can set `odbc_eof` as a member of this odbc scanner. and the `eof` can be set to true iff:
1. `odbc_eof` is true, and
2. block->rows() == 0

And each time we call `_get_block_impl()`, we check `odbc_eof` first, if it is true, no need to read data and return with `eof` as true.